miner/worker.go

```diff
@@ -572,50 +573,95 @@ func (w *worker) taskLoop() {
 
 		prevParentHash common.Hash
 		prevProfit     *big.Int
+		timer          = time.NewTimer(0)
 	)
 
+	<-timer.C // usual skip first tick
+
 	// interrupt aborts the in-flight sealing task.
 	interrupt := func() {
 		if stopCh != nil {
 			close(stopCh)
 			stopCh = nil
 		}
 	}
+
+	taskWork := map[string]*task{}
+	var defaultTaskPick *task
+
+	handle := func(task *task) {
+		if w.newTaskHook != nil {
+			w.newTaskHook(task)
+		}
+		// Reject duplicate sealing work due to resubmitting.
+		sealHash := w.engine.SealHash(task.block.Header())
+		if sealHash == prev {
+			return
+		}
+
+		taskParentHash := task.block.Header().ParentHash
+		// reject new tasks which don't profit
+		if taskParentHash == prevParentHash &&
+			prevProfit != nil && task.profit.Cmp(prevProfit) < 0 {
+			return
+		}
+		prevParentHash = taskParentHash
+		prevProfit = task.profit
+
+		// Interrupt previous sealing operation
+		interrupt()
+		stopCh, prev = make(chan struct{}), sealHash
+		log.Info("Proposed miner block", "blockNumber", task.block.Number(), "profit", prevProfit, "isFlashbots", task.isFlashbots, "sealhash", sealHash, "parentHash", prevParentHash)
+		if w.skipSealHook != nil && w.skipSealHook(task) {
+			return
+		}
+		w.pendingMu.Lock()
+		w.pendingTasks[sealHash] = task
+		w.pendingMu.Unlock()
+
+		if err := w.engine.Seal(w.chain, task.block, w.resultCh, stopCh); err != nil {
+			log.Warn("Block sealing failed", "err", err)
+		}
+
+	}
+
 	for {
 		select {
-		case task := <-w.taskCh:
-			if w.newTaskHook != nil {
-				w.newTaskHook(task)
-			}
-			// Reject duplicate sealing work due to resubmitting.
-			sealHash := w.engine.SealHash(task.block.Header())
-			if sealHash == prev {
-				continue
-			}
 
-			taskParentHash := task.block.Header().ParentHash
-			// reject new tasks which don't profit
-			if taskParentHash == prevParentHash &&
-				prevProfit != nil && task.profit.Cmp(prevProfit) < 0 {
-				continue
-			}
-			prevParentHash = taskParentHash
-			prevProfit = task.profit
+		case <-timer.C:
+			// timer went off, use first bundle, defaultTaskPick
+			handle(defaultTaskPick)
```

Comment:
possible null pointer if no task has ever been sent to taskCh.

Also, related, what is the intended logic if neither the vanilla worker or the flashbots worker has produced a task within the time limit.

Reply:
Never mind, forgot this channel is only triggered after a task is received.

Reply:
not a possible/relevant situation? (not sure) Because the timer can only go off if there has been at least one task right? - I'll add the nil pointer check, but it won't be possible i think

Reply:
ah - ya, I see you wrote, my github didn't update with the second comment